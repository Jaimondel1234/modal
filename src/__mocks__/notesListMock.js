export const notesListMock = [
  {
    id: 125465,
    title: 'first element',
    description: 'description first element',
    imgUrl: '',
  },
  {
    id: 125466,
    title: 'second element',
    description: 'description second element',
    imgUrl: '',
  },
  {
    id: 125467,
    title: 'third element',
    description: 'description third element',
    imgUrl: '',
  },
];
