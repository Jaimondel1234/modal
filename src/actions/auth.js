import { types } from "../types/types";

export const authLogin = (email, password) => ({
  type: types.authLogin,
  email,
  password,
});

export const authLogout = () => ({
  type: types.authLogout,
});
