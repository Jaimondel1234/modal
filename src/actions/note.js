import { types } from "../types/types";

export const noteAddNew = (newNotes) => ({
  type: types.noteAddNew,
  newNotes,
});

export const noteDelete = (noteId) => ({
  type: types.noteDeleted,
  noteId,
});

export const noteSetActive = (activeNote) => ({
  type: types.noteSetActive,
  activeNote,
});

export const noteUpdate = (note) => ({
  type: types.noteUpdate,
  note,
});

export const setListMode = (mode) => ({
  type: types.noteSetListMode,
  mode,
});

export const noteReset = () => ({
  type: types.noteReset,
});
