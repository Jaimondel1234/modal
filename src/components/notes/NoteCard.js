import React, { useState } from "react";
import {
  PanelStyle,
  ImageStyle,
  TextStyle,
  ImgText,
  ImgContainer,
} from "../../styles/core";
import PropTypes from "prop-types";
import { DeleteEventFab } from "../ui/DeleteEventFab";
import { ModifyNoteFab } from "../ui/ModifyNoteFab";

/*
 Card with the information of a note
*/
export const NoteCard = ({ ...note }) => {
  const [isMouveIn, setIsMouseIn] = useState(false);

  const { id, title, imgUrl, description } = note;

  const handleMouseOver = () => {
    setIsMouseIn(true);
  };

  const handleMouseOut = () => {
    setIsMouseIn(false);
  };

  return (
    <PanelStyle
      ml='20px'
      mr='20px'
      width='450px'
      mb='50px'
      borderRadius='10px'
      onMouseOver={handleMouseOver}
      onMouseLeave={handleMouseOut}
    >
      <ImgContainer>
        <ImageStyle
          src={imgUrl}
          width='450px'
          height='400px'
          title={title}
          onError={(e) => {
            e.target.onerror = null;
            e.target.src =
              "https://www.escapadarural.com/blog/wp-content/uploads/2019/08/Playa-de-la-Tejita.-Tenerife.jpg";
          }}
        />
        {isMouveIn && <DeleteEventFab noteId={id} />}
        {isMouveIn && <ModifyNoteFab note={note} />}
        <ImgText color='white' fontSize='30px' fontWeight='bold'>
          {title}
        </ImgText>
      </ImgContainer>
      <TextStyle mb='30px' mr='25px' ml='30px' fontSize='25px'>
        {description}
      </TextStyle>
    </PanelStyle>
  );
};

NoteCard.propTypes = {
  note: PropTypes.shape({
    id: PropTypes.number,
    title: PropTypes.string,
    imgUrl: PropTypes.string,
    description: PropTypes.string,
  }),
};
