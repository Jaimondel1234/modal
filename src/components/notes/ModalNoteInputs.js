import React from "react";
import Swal from "sweetalert2";
import { noteAddNew, noteUpdate, noteSetActive } from "../../actions/note";
import { useForm } from "../../hooks/useForm";
import styled from "styled-components";
import { InputStyle, FlexStyle, ButtonStyle } from "../../styles/core";
import { useDispatch, useSelector } from "react-redux";
import { uiCloseModal } from "../../actions/ui";

/** Style of modal input */
const ModalInputStyle = styled(InputStyle)`
  border-radius: 7px;
  font-size: 30px;
  margin-bottom: 20px;
  width: 95%;
  padding: 15px;
  padding-left: 25px;
`;

/*
 Modal Note Input
*/
export const ModalNoteInputs = () => {
  const dispatch = useDispatch();
  const { activeNote } = useSelector((state) => state.notes);
  const defaultEvent = {
    title: "",
    description: "",
    imgUrl: "",
  };

  const initEvent = activeNote || defaultEvent;
  const [formValues, handleInputChange, reset] = useForm(initEvent);

  const { description, title, imgUrl } = formValues;

  const closeModal = () => {
    dispatch(uiCloseModal());

    if (activeNote) {
      dispatch(noteUpdate(formValues));
      dispatch(noteSetActive(false));
    } else {
      const noteWithId = { id: new Date().getTime(), ...formValues };
      dispatch(noteAddNew([noteWithId]));
    }

    reset(initEvent);
  };
  const handleSubmitForm = (e) => {
    e.preventDefault();
    if (title.trim().length === 0) {
      Swal.fire("The title is required", "", "error");
    } else if (description.trim().length === 0) {
      Swal.fire("The description is required", "", "error");
    } else {
      Swal.fire("Note Saved", "", "success");
      closeModal();
    }
  };
  return (
    <FlexStyle
      as='form'
      alignItems='center'
      flexDirection='column'
      onSubmit={handleSubmitForm}
    >
      <ModalInputStyle
        type='text'
        placeholder='Title'
        name='title'
        autoComplete='off'
        value={title}
        onChange={handleInputChange}
        mt='30px'
      />

      <ModalInputStyle
        as='textarea'
        type='text'
        placeholder='Description'
        rows='5'
        name='description'
        value={description}
        onChange={handleInputChange}
        p='10px'
      />

      <ModalInputStyle
        type='text'
        placeholder='Imagen (Url)'
        name='imgUrl'
        autoComplete='off'
        value={imgUrl}
        onChange={handleInputChange}
      />

      <ButtonStyle
        p='15px'
        fontSize='30px'
        fontWeight='bold'
        letterSpacing='3px'
        lineHeight='40px'
        width='100%'
        type='submit'
      >
        Save
      </ButtonStyle>
    </FlexStyle>
  );
};
