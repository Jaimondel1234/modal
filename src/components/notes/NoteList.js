import React from "react";
import { FlexStyle } from "../../styles/core";
import map from "lodash/map";
import { NoteCard } from "./NoteCard";
import { useSelector } from "react-redux";
import { getListByMode } from "../../helpers/helpers";

/*
 Cards List
*/
export const NoteList = () => {
  const { notes, listMode } = useSelector((state) => state.notes);
  let noteListToShow = getListByMode(listMode, notes);

  return (
    <FlexStyle
      alignItem='center'
      justifyContent='center'
      flexDirection='column'
      mt='30px'
    >
      <FlexStyle
        alignItem='center'
        justifyContent='center'
        flexDirection='row'
        flexWrap='wrap'
      >
        {noteListToShow &&
          noteListToShow.length > 0 &&
          map(noteListToShow, (props) => (
            <NoteCard key={props.id} {...props} />
          ))}
      </FlexStyle>
    </FlexStyle>
  );
};
