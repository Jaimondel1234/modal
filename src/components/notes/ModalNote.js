import React from 'react';
import Modal from 'react-modal';
import { TextStyle } from '../../styles/core';
import { useSelector } from 'react-redux';
import { ModalNoteInputs } from './ModalNoteInputs';

const customStyles = {
  content: {
    top: '50%',
    left: '50%',
    right: 'auto',
    bottom: 'auto',
    marginRight: '-50%',
    transform: 'translate(-50%, -50%)',
    width: '50vw',
    padding: '50px',
  },
};

/**
 * Component of modal to create or update a note
 */
export const ModalNote = () => {
  if (process.env.NODE_ENV !== 'test') {
    Modal.setAppElement('#root');
  }
  const { modalOpen } = useSelector((state) => state.ui);
  const { activeNote } = useSelector((state) => state.notes);

  return (
    <Modal
      isOpen={modalOpen}
      style={customStyles}
      closeTimeoutMS={200}
      ariaHideApp={!(process.env.NODE_ENV === 'test')}
    >
      <TextStyle fontSize='60px' fontWeight='bold' textAlign='center'>
        {activeNote ? 'Change your Note' : 'Create your Note'}
      </TextStyle>
      <hr />
      <ModalNoteInputs />
    </Modal>
  );
};
