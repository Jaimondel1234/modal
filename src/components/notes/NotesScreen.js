import React, { useEffect } from "react";
import { NoteList } from "./NoteList";
import { ContainerStyle } from "../../styles/core";
import { AddNewFab } from "../ui/AddNewFab";
import { ModalNote } from "./ModalNote";
import { useSelector } from "react-redux";
import { ButtonsSelector } from "./ButtonsSelector";

/*
  Component of Screen Notes
*/
export const NotesScreen = () => {
  const { notes } = useSelector((state) => state.notes);

  useEffect(() => {
    localStorage.setItem("notes", JSON.stringify(notes));
  }, [notes]);

  const { modalOpen } = useSelector((state) => state.ui);
  return (
    <>
      <ContainerStyle mt='20px'>
        <ButtonsSelector />
        <NoteList />
        <AddNewFab />
      </ContainerStyle>
      {modalOpen && <ModalNote />}
    </>
  );
};
