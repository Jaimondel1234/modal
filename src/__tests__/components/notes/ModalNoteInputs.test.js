import React from 'react';
import { notesListMock } from '../../../__mocks__/notesListMock';
import { mount } from 'enzyme';
import { act } from 'react-dom/test-utils';
import configureStore from 'redux-mock-store';
import { Provider } from 'react-redux';
import { ModalNoteInputs } from '../../../components/notes/ModalNoteInputs';
import { uiCloseModal } from '../../../actions/ui';
import { noteSetActive, noteUpdate, noteAddNew } from '../../../actions/note';
import Swal from 'sweetalert2';

jest.mock('../../../actions/note', () => ({
  noteUpdate: jest.fn(),
  noteSetActive: jest.fn(),
  noteAddNew: jest.fn(),
}));
jest.mock('../../../actions/ui', () => ({
  uiCloseModal: jest.fn(),
}));
jest.mock('sweetalert2', () => ({
  fire: jest.fn(),
}));
describe('Test <ModalNoteInputs/>', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });
  const mockStore = configureStore([]);
  const store = mockStore({
    notes: { activeNote: null },
  });
  store.dispatch = jest.fn();
  const wrapper = mount(
    <Provider store={store}>
      <ModalNoteInputs />
    </Provider>,
  );
  test('it must be displayed correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });
  test('if a new note is created, the actions must be correct', () => {
    wrapper.find('input[name="title"]').simulate('change', {
      target: {
        name: 'title',
        value: 'the title',
      },
    });
    wrapper.find('textarea[name="description"]').simulate('change', {
      target: {
        name: 'description',
        value: 'the decription',
      },
    });

    act(() => {
      wrapper.find('form').at(0).prop('onSubmit')({
        preventDefault() {},
      });
      expect(Swal.fire).toHaveBeenCalledWith('Note Saved', '', 'success');
      expect(uiCloseModal).toHaveBeenCalled();
      expect(noteAddNew).toHaveBeenCalledWith([
        {
          id: expect.any(Number),
          title: 'the title',
          description: 'the decription',
          imgUrl: '',
        },
      ]);
    });
  });
  test('The title input is required', () => {
    wrapper.find('input[name="title"]').simulate('change', {
      target: {
        name: 'title',
        value: '',
      },
    });

    act(() => {
      wrapper.find('form').at(0).prop('onSubmit')({
        preventDefault() {},
      });
      expect(Swal.fire).toHaveBeenCalledWith(
        'The title is required',
        '',
        'error',
      );
    });
  });
  test('The description input is required', () => {
    wrapper.find('input[name="title"]').simulate('change', {
      target: {
        name: 'title',
        value: 'the title',
      },
    });
    wrapper.find('textarea[name="description"]').simulate('change', {
      target: {
        name: 'description',
        value: '',
      },
    });
    act(() => {
      wrapper.find('form').at(0).prop('onSubmit')({
        preventDefault() {},
      });
      expect(Swal.fire).toHaveBeenCalledWith(
        'The description is required',
        '',
        'error',
      );
    });
  });

  test('if there is an active note, the actions dispatched must be correct', () => {
    const mockStore = configureStore([]);
    const store = mockStore({
      notes: { activeNote: notesListMock[0] },
    });
    store.dispatch = jest.fn();
    const wrapper = mount(
      <Provider store={store}>
        <ModalNoteInputs />
      </Provider>,
    );
    wrapper.find('input[name="title"]').simulate('change', {
      target: {
        name: 'title',
        value: 'the title',
      },
    });

    act(() => {
      wrapper.find('form').at(0).prop('onSubmit')({
        preventDefault() {},
      });
    });
    expect(uiCloseModal).toHaveBeenCalled();
    expect(noteUpdate).toHaveBeenCalledWith({
      ...notesListMock[0],
      title: 'the title',
    });
    expect(noteSetActive).toHaveBeenCalledWith(false);
    expect(Swal.fire).toHaveBeenCalledWith('Note Saved', '', 'success');
  });
});
