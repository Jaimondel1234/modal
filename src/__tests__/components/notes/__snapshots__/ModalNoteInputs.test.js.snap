// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test <ModalNoteInputs/> it must be displayed correctly 1`] = `
<form
  className="core__FlexStyle-sc-1mkz20q-3 iTHhnD"
  onSubmit={[Function]}
>
  <input
    autoComplete="off"
    className="core__InputStyle-sc-1mkz20q-4 ModalNoteInputs__ModalInputStyle-sc-18t1odo-0 bMnAXg ksglXK"
    name="title"
    onChange={[Function]}
    placeholder="Title"
    type="text"
    value=""
  />
  <textarea
    className="core__InputStyle-sc-1mkz20q-4 ModalNoteInputs__ModalInputStyle-sc-18t1odo-0 fsMNMP ksglXK"
    name="description"
    onChange={[Function]}
    placeholder="Description"
    rows="5"
    type="text"
    value=""
  />
  <input
    autoComplete="off"
    className="core__InputStyle-sc-1mkz20q-4 ModalNoteInputs__ModalInputStyle-sc-18t1odo-0 kivKiO ksglXK"
    name="imgUrl"
    onChange={[Function]}
    placeholder="Imagen (Url)"
    type="text"
    value=""
  />
  <button
    className="core__ButtonStyle-sc-1mkz20q-5 cgFcMK"
    fontSize="30px"
    fontWeight="bold"
    letterSpacing="3px"
    type="submit"
    width="100%"
  >
    Save
  </button>
</form>
`;
